### PR TITLE
syncthing-inotify 0.6.5 (new formula)

### PR DESCRIPTION
**Pull request by Kerosene2000**
A simple add but really useful I think, if you want [Syncthing](https://github.com/syncthing/syncthing) to become a real-time indexer !

### CODE REVIEW

File: Library/Formula/syncthing-inotify.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    :tag => "v0.6.5", :revision => "2b290b929e18e7cd1cfea11240e4fa870d6638d5"
+
+  head "https://github.com/syncthing/syncthing-inotify.git"
+
+  bottle do
+    cellar :any
+    sha256 "392d1b861c5eae39e44148842947dfbeb3640cb0d031e0a42fef3ec49d31252b" => :yosemite
+    sha256 "28b78c3826f3a9a6c1b521ad91b2e7c91b3bc6417f7e175a2e83960ad2f78e1d" => :mavericks
+    sha256 "a1c9ae1947ae60898b14820577ffafc4bbb5e92039d02ebf9e8d2aa9e5fffaea" => :mountain_lion
+  end
```
Comment (mikemcquaid) [inline Library/Formula/syncthing-inotify.rb]: Remove this; not sure where you got these values for but they will need to be built by us and added after this formula is included.

File: Library/Formula/syncthing-inotify.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    ENV["GOPATH"] = cached_download/".gopath"
+    ENV.append_path "PATH", "#{ENV["GOPATH"]}/bin"
+
+    # FIXTHIS: do this without mutating the cache!
+    hack_dir = cached_download/".gopath/src/github.com/syncthing"
+    rm_rf hack_dir
+    mkdir_p hack_dir
+    ln_s cached_download, "#{hack_dir}/syncthing-inotify"
+
+    system "go", "get", "-d"
```
Comment (mikemcquaid) [inline Library/Formula/syncthing-inotify.rb]: No go dependencies?

Comment (bfontaine) [inline Library/Formula/syncthing-inotify.rb]: and we should use `go_resource` instead of `go get` (which will download external stuff).

File: Library/Formula/syncthing-inotify.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
+class SyncthingInotify < Formula
+  desc "File watcher intended for use with Syncthing"
+  homepage "https://github.com/syncthing/syncthing-inotify"
+  url "https://github.com/syncthing/syncthing-inotify.git",
+    :tag => "v0.6.5", :revision => "2b290b929e18e7cd1cfea11240e4fa870d6638d5"
```
Comment (bfontaine) [inline Library/Formula/syncthing-inotify.rb]: Why not using a tarball? https://github.com/syncthing/syncthing-inotify/archive/v0.6.5.tar.gz

File: Library/Formula/syncthing-inotify.rb (hunk 2 above)
Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: did I just need to change `system "go", "get", "-d"` by `go_resource` ?

File: Library/Formula/syncthing-inotify.rb (hunk 3 above)
Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: when I change by this:
```ruby
  url "https://github.com/syncthing/syncthing-inotify/archive/v0.6.5.tar.gz"
  version "0.6.5"
  sha256 "430297896bb05396268fd29cc555eba6542b42263489784c9843f4daf625ac5c"
```
I get an error:
```
==> Downloading https://github.com/syncthing/syncthing-inotify/archive/v0.6.5.tar.gz
==> Downloading from https://codeload.github.com/syncthing/syncthing-inotify/tar.gz/v0.6.
######################################################################## 100,0%
Error: File exists - /Library/Caches/Homebrew/syncthing-inotify-0.6.5.tar.gz
```

It seems it download multiple times the same tarball...

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: No in fact, there is no multiple download but homebrew is confused...

**Comment by Kerosene2000** [comment]
I am working on the bug `Error: File Exist` that's me sorry

**Comment by Kerosene2000** [comment]
Can you explain to me this error please:
`GitHub fork (not canonical repository)`
Is this because [syncthing/syncthing-inotify](https://github.com/syncthing/syncthing-inotify) is a fork ?

File: Library/Formula/syncthing-inotify.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  desc "File watcher intended for use with Syncthing"
+  homepage "https://github.com/syncthing/syncthing-inotify"
+  url "https://github.com/syncthing/syncthing-inotify/archive/v0.6.5.tar.gz"
+  sha256 "430297896bb05396268fd29cc555eba6542b42263489784c9843f4daf625ac5c"
+
+  depends_on "go" => :build
+
+  go_resource "github.com/cenkalti/backoff" do
+    url "https://github.com/cenkalti/backoff.git",
+        :revision => "6c45d6bc1e78d94431dff8fc28a99f20bafa355a" # not sure !
```
Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: Usually we just use the last commit, as long as it builds.

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: Ok, how can I fix this error ?
Using a different commit ?

Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: Oh, this is unrelated to the `GitHub fork (not canonical repository)` error.  You're right that we get that error because this is a fork (using a fork is a good, but not perfect, indicator of a non-official repository).

My comment here was just about your `not sure!` comment.  If you're not sure what revision to use, just grab the most recent.

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: Ho ok ! That's what I have done, thank you !
It's fun y because this is the official Syncthing team which fork an unofficial extension abd don't pull request, then everyone use the fork in case of the old syncthing-inotify repo ;)

Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: Yeah, I think in this case we can probably ignore that warning!

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: Ok I will ask the validation team to take a look at those two repos and probably make an exception.
Thanks

**Comment by Kerosene2000** [comment]
Hello validation team !

I have a little problem, I want to push this `Formula` but it's based on a `fork repo`.
If you take a look at [the canonical repo](https://github.com/a-m-s/syncthing-inotify) and [the fork repo by the syncthing team](https://github.com/syncthing/syncthing-inotify) you can see that the second is really better and maintained than the first...

Is there a possibility to make an exception related to the `GitHub fork (not canonical repository)` error ?

**Comment by mikemcquaid** [comment]
@Kerosene2000 It's fine, we can merge as-is!

File: Library/Formula/syncthing-inotify.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        <string>#{var}/log/syncthing-inotify.log</string>
+        <key>StandardOutPath</key>
+        <string>#{var}/log/syncthing-inotify.log</string>
+      </dict>
+    </plist>
+    EOS
+  end
+
+  test do
+    system "syncthing-inotify", "-version"
```
Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: Is there any chance of a more involved test, though? It'd be great to test a bit more of the functionality of the program.

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: I am ok with you but what can I had ?

```
Options:
  -api="7S03Mx30rUX6IAltJ2dU0f0HCVoJtRPD"  API key
  -api-stdin                               Provide API key through stdin
  -csrf=""                                 CSRF token file
  -delay-scan="86400"                      Automatically delay next scan interval (in seconds)
  -folders=""                              A comma-separated list of folders to watch (all by default)
  -home=""                                 Specify the home Syncthing dir to sniff configuration settings
  -logflags="2"                            Select information in log line prefix
  -password="***"                          Password
  -password-stdin                          Provide password through stdin
  -skip-folders=""                         A comma-separated list of folders to skip inotify watching
  -target="http://127.0.0.1:8384"          Target url (prepend with https:// for TLS)
  -user=""                                 Username
  -verbosity="2"                           Logging level [1..4]
  -version                                 Show version


The -logflags value is a sum of the following:

   1  Date
   2  Time
   4  Microsecond time
   8  Long filename
  16  Short filename

I.e. to prefix each log line with date and time, set -logflags=3 (1 + 2 from
above). The value 0 is used to disable all of the above. The default is to
show time only (2).
```

Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: Maybe try running it for a few seconds with `fork do` (see [`entr.eb`](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/entr.rb) for an example), then see if log files are created correctly?

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: when did this test is called ?
At the end of the installation ?
Because I'am testing the installation on my own computer and I didn't see my logs tests...

Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: You can run the test with `brew test syncthing-inotify`.  The test is run in a temporary directory, so any log files will be deleted after the test is run.

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: but in fact, this is the commands:
`syncthing-inotify 2> #{var}/log/syncthing-inotify.log`
`syncthing-inotify 1> #{var}/log/syncthing-inotify.log`
which create the log file with the `.plist` file.

if want to test if the `.plist` works correctly I need (cf. `brew install syncthing-inotify`):
```
To have launchd start syncthing-inotify at login:
    ln -sfv /usr/local/opt/syncthing-inotify/*.plist ~/Library/LaunchAgents
Then to load syncthing-inotify now:
    launchctl load ~/Library/LaunchAgents/homebrew.mxcl.syncthing-inotify.plist
```

Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: That's to have it start on login, but you don't need launchctl to run it in general.

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: Then how do you want me to test the log file existance if I can't use `.plist file` ?

Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: Can you run manually the command specified in the `.plist` file?

Comment (Kerosene2000) [inline Library/Formula/syncthing-inotify.rb]: I think, but there is no interest in doing this...
`syncthing-inotify 2> #{var}/log/syncthing-inotify.log`
if this command failed, it will return `$? != 0` then it's the same than just launching `syncthing-inotify` without other stuff

Comment (dunn) [inline Library/Formula/syncthing-inotify.rb]: Alright, I guess we can take the test as-is. Thanks!

**Comment by dunn** [comment]
Merged as ddffe51ca97e21d0625389a3ac03bfae10f48dd4, thanks again for the submission!

**Comment by Kerosene2000** [comment]
Thanks to all the validation team for all the help !
I love this project ;)
